Add a couple new cases to DateTimeFormat::fix()

MrPetovan opened the pull request:
Re-fix #11938

- Reworked method to perform more string replacements and fewer regular expression matches

MrPetovan committed:
Only call DateTimeFormat::fix() when there's an Exception

- This prevents valid date/time strings to be mangled by fix()

annando commented on src/Util/DateTimeFormat.php:
```diff
@@ -184,17 +184,17 @@ public static function convert($s = 'now', $tz_to = 'UTC', $tz_from = 'UTC', $fo
-			['#(\w+), (\d+/\d+/\d+) - (\d+:\d+)#', '$1, $2 $3'],
-			['#(\d+-\d+-\d+)T(\d+:\d+:\d+)ZZ#', '$1T$2Z'],
-			['#(\d+-\d+-\d+)T(\d+:\d+:\d+\.\d+)ZZ#', '$1T$2Z'],
+		$search  = ['Mär', 'März', 'Mai', 'Juni', 'Juli', 'Okt', 'Dez', 'ET' , 'ZZ', ' - ', '&#x2B;'];
```
We mustn't use these simple patterns to fix the problem. We are not only calling this function with fixed date values but also with date formulas. When we replace " - " with "" then a date formula like "now - 3 days" will stop working. I accidentally deleted many posts this way some weeks ago :grin:

MrPetovan replied:
This is where function comments are helping in a big way. As long as this kind of knowledge stays in your head, it can only help yourself. As soon as you put it down for others, it helps everybody save their time.

nupplaphil replied:
And adding each edge case a test so we can avoid such regressions ;-)))

annando replied:
Problem is that it is not really an edge case. It is based on the way that functionality works. This functionality is called both with static values and with formulas. And thus problems like these can always occur.

MrPetovan commented:
@annando I have added the relative date string to the test dataset, tweaked the method and we're good, unless you know about another corner case that we should add to the tests as well?

annando commented:
I only stumbled across this ` - ` thing. But I just had an idea. What if we only apply the fix when the first call of `$d = new DateTime($s, $from_obj);` fails? Doing so, we will avoid any side effects, I guess.

MrPetovan commented:
This is a good idea.

MrPetovan committed:
Add a couple of cases to DateTimeFormat::fix()

- Reworked method to perform more string replacements and fewer regular expression matches

MrPetovan commented:
This feels slightly bad to me but I'm glad we saved hundreds of unwarranted calls to `DateTimeFormat::fix()`. Still, I've added a canary test to make sure that relative date/time strings still work as intended given the mangling risk.
